smart_open/s3.py

```diff
@@ -738,8 +737,7 @@ def read(self, size=-1):
 
             if binary:
                 self._position += len(binary)
-                binary_collected.append(binary)
-                bytes_read += len(binary)
+                binary_collected += binary
```

Comment:
Not a fan. The previous version was more explicit & simpler IMO.

Is it really guaranteed `bytearray +=` is efficient and doesn't keep re-allocating internally?

Reply:
So, logarithmic reallocations. Not as good as the original `join()` which is constant, but not terrible I guess.

Reply:
according to [this article](https://pythonspeed.com/articles/bytesio-reduce-memory-usage/), our best bet seems to be `BytesIO.get_value()`